zrml/prediction-markets/src/migrations.rs

```diff
@@ -337,6 +349,99 @@ impl<T: Config> OnRuntimeUpgrade for CleanUpStorageForResolvedOrClosedMarkets<T>
     }
 }
 
+impl<T: Config> OnRuntimeUpgrade for MigrateMarketIdsPerBlockStorage<T> {
+    fn on_runtime_upgrade() -> Weight
+    where
+        T: Config,
+    {
+        let mut total_weight = T::DbWeight::get().reads(1);
+
+        if StorageVersion::get::<Pallet<T>>()
+            != PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION_FOR_MIGRATE_MARKET_IDS_STORAGE
+            || utility::get_on_chain_storage_version_of_market_commons_pallet()
+                != MARKET_COMMONS_REQUIRED_STORAGE_VERSION
+        {
+            log::info!(
+                "Skipping storage migration for MarketIds; prediction-markets already up to date"
+            );
+            return total_weight;
+        }
+
+        total_weight = total_weight.saturating_add(T::DbWeight::get().reads(2));
+        log::info!("Starting storage cleanup of MigrateMarketIdsPerBlockStorage");
+
+        type DisputeBlockToMarketIdsTuple<T> =
+            (<T as frame_system::Config>::BlockNumber, BoundedVec<MarketIdOf<T>, CacheSize>);
+        type IterType<T> = PrefixIterator<DisputeBlockToMarketIdsTuple<T>>;
+
+        let market_ids_per_dispute_iterator: IterType<T> =
+            frame_support::migration::storage_key_iter::<_, _, Twox64Concat>(
+                b"PredictionMarkets",
+                b"MarketIdsPerDisputeBlock",
+            );
+
+        let market_ids_per_dispute: Vec<DisputeBlockToMarketIdsTuple<T>> =
+            market_ids_per_dispute_iterator.collect();
+        for (dispute_start_block, market_ids) in market_ids_per_dispute {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(2));
+            // NOTE: These migration only makes sense on BS runtime so its fine to assume
+            // DisputePeriod equal to BLOCKS_PER_DAY
+            let dispute_period: T::BlockNumber = BLOCKS_PER_DAY.saturated_into::<u32>().into();
+            let new_dispute_start_block = dispute_start_block.saturating_add(dispute_period);
+            MarketIdsPerDisputeBlock::<T>::insert(new_dispute_start_block, market_ids);
+            MarketIdsPerDisputeBlock::<T>::remove(dispute_start_block);
+        }
+
+        let market_ids_per_report_iterator: IterType<T> =
+            frame_support::migration::storage_key_iter::<_, _, Twox64Concat>(
+                b"PredictionMarkets",
+                b"MarketIdsPerReportBlock",
+            );
+
+        let market_ids_per_report: Vec<_> = market_ids_per_report_iterator.collect();
+        for (dispute_start_block, market_ids) in market_ids_per_report {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(2));
+            // NOTE: These migration only makes sense on BS runtime so its fine to assume
+            // DisputePeriod equal to BLOCKS_PER_DAY
```

Comment:
As above, unfortunately, by the time this migration is executed, there will be probably be markets on Zeitgeist.